ycmd/tests/get_completions_test.py

```diff
@@ -115,9 +123,800 @@ def GetCompletions_IdentifierCompleter_WorksForSpecialIdentifierChars_test():
                            completion_data ).json[ 'completions' ]
 
   assert_that( results,
-               has_items( CompletionEntryMatcher( 'font-size' ),
-                          CompletionEntryMatcher( 'font-family' ) ) )
+               has_items( CompletionEntryMatcher( 'font-size', '[ID]' ),
+                          CompletionEntryMatcher( 'font-family', '[ID]' ) ) )
+
 
+@with_setup( Setup )
+def GetCompletions_Fall_Back_ClangCompleter_test():
+
+  # because we use it a lot
+  NO_COMPLETIONS_ERROR = ErrorMatcher( RuntimeError, NO_COMPLETIONS_MESSAGE )
+
+  """
+  GetCompletions_Fall_Back_test
+
+  Tests for:
+    - [X] multiple languages: (use cases)
+      - [X] C:      macros
+      - [X] C++:    templates
+      - [X] Python: properties set from external method
+    - [X] Forced vs. Triggered semantic completion
+    - [X] Identifiers tagged with '[ID]'
+    - [X] Filter/sort based on query
+    - [X] Min chars for completion (only applies to ID)
+    - [X] Extra conf throws exception
```

Comment:
That perspective is exemplary but in my experience leads to tests that can be hard to maintain and a general test suite that isn't as helpful as it could be. I've worked in a codebase with tons of tests all testing the same thing indirectly, and when one such important component failed, it was a nightmare to figure out what actually went wrong. You'd know _something_ went south because 100+ tests failed, but good luck figuring it out.

Since then I've focused on the mantra that "all functionality should be tested while trying to minimize test path duplication."

Think of it like a compiler error: if you forget a semicolon, would you rather see a spew of 100+ errors or just one saying "hey you forgot a semicolon here"? C++ for instance has compilers that spew out an ungodly mess of text on incorrect template usage and figuring out what exactly went wrong is hard.

> my perspective has always been that it's really useful to write regression tests (in particular integration tests testing an interface - as these are) from the functional perspective, so that if the code is changed (arbitrarily) then the tests should still work.

Black-box testing can be very useful, but isolated unit-tests IMO are more useful. But the principle you're espousing of "tests should still work if the code is changed" is absolutely something that's critical and IMO not at odds with my approach.